### PR TITLE
Reenable updates

### PR DESCRIPTION
**Pull request by pkegg**
**tl;dr:** Updates work as before. `daily` update channel is now called `release`.  There are new developer settings in ES to switch from downloading releases from 351ELEC github org to `pkegg` (or any other github org) for testing.

**to test**: 0. Install this build. 1. Switch `EMULATIONSTATION SETTINGS -> DEVELOPER -> GITHUB ORG` to `pkegg` (no releases on 351ELEC to upgrade to)  2. `UPDATES & DOWNLOADS -> START UPDATE`  3. (cleanup) Remove `pkegg` entry from step 1.

## Summary

This brings back the `Downloads & Updates` menu in ES (instead of ‘Downloads’) that can be used to do an 'online' update.  Using online update is functionally identical to manually placing the tar into roms/updates - except you can trigger the download from ES on your device.

The 'Check for Updates` flag/popup also works and displays a popup when a new release is found.  Updates are now  downloaded directly from github releases, so some developer options are added to switch the github org/repository for testing.

**Changes**
- Re-enable updates and the automatic check for updates.
  - Call the default update channel `release` instead of `daily`
  - Add a `beta` channel which will pull releases marked `prerelease` in github.  This will allow 351ELEC team to publish a 'release candidate' build just by marking a release as 'prerelease`.  
    - `beta` channel will also get 'releases' (for example, if the newest release is a normal release and not a pre-release, `beta` will still use the normal release and not an older pre-release).
    - **Release Process**: Github allows switching the `prerelease` flag after creating a release.  This means, 351ELEC can either issue an explicit 'release candidate' build and then create a *new* build for a release **or** it can issue a release marked `prerelease` and remove the flag if that should become the 'release'.  This decision is TBD.

- **Testing Info**: To test with a *real* release, there's a new developer option in ES to switch your github org to 'pkegg' (see Testing on the RG351 section below).
  -  FYI: the pkegg updates are releases are from this PR, so you can keep updating over and over, etc. They are 'tagged' as newer, but internally they are marked 20210603, so they will keep updating for testing purposes.

### Future-Proofing
One issue with putting an updater into software is we need to plan for backwards incompatible changes in the future and how they will interact with updates.  For example: we have an update which *requires* reflashing or we end up changing the version numbering scheme.

In order to deal with this future proofing, this change leans into the version numbering schema.  Here is the logic:
- The first github release (ordered by newest releases first) which matches our version numbering (`yyyymmdd`) will be used.  Ex: `20210603`.  This means if we end up publishing a version like 99991212, we can just publish a newer release with the correct numbering to supercede it.  However, anyone on the 99991212 build will not be able to do an online update as it won't be 'newer'

  - **Patch Versioning** - Also support a versioning of `20210603-1` to allow multiple builds on one day w/o reusing the same identifier.
- Versions which don't match the above will be ignored.

This means, if we need to *force* an update to a specific version and nothing newer, we can switch the naming convention and it will only update to the latest version with this naming convention (For example, we could prepend version with `v`, like `v20210603` or whatever).  

If we change the versioning numbering for unrelated reasons, we can provide an update with this numbering that can deal with the new numbering.

## Technical Details
- Reimplemented the bulk of the update logic in python3 as opposed to bash.  This just makes it much easier to deal with the github API (parse json) and allows testing when running on Ubuntu or other linux systems that are not the RG351.
  - Also dealt with the github API limitations - if more than 60 update checks are done per hour, github will throttle.  The script should indicate how long you need to wait before updates can be checked for again.  This should be a rare occurrence - I only hit it once while developing this. 
- Kept the `/usr/bin/updatecheck` and `/usr/bin/351elec-upgrade` scripts for compatibility with the rest of the system and to have the device specific code (`systemctl stop emustation`, `restart`, etc)

## Development Testing
This section refers for how to debug/update the main update script itself from a development environment (another advantage of using python).  See the following section for testing this PR on the device.

- Typically for dev testing, you would directly use the `/usr/bin/get-release.py` (in source: `packages/351elec/sources/scripts/get-release.py`) command to check for updates and test downloads.  The script provides it's own help when run with `-h`.  Ex:
```
  -h, --help            show this help message and exit
  --org ORG             Github organization. Allows testing with fork other than 351ELEC
  --band {release,beta,daily}
                        Update "band" ("channel"). Allows determining what latest release to get. "daily" is for backwards compatibility and maps to "release"
  --device {RG351P,RG351V}
                        Sets the appropriate device for testing. Will fallback to contents of /storage/.config/.OS_ARCH
  --console CONSOLE     Sets device to output messages to. Use /dev/stderr for testing.
  --log-level {DEBUG,INFO,WARN,ERROR,CRITICAL}, -l {DEBUG,INFO,WARN,ERROR,CRITICAL}
                        Allows setting log level for more debugging
  --update-dir UPDATE_DIR
                        Allows setting a different directory for downloads
  --existing-release EXISTING_RELEASE
                        Overrides release version in "/storage/.config/.OS_VERSION for upgrade/testing purposes
  --check CHECK         Only check for update - do not download update
  --force-update FORCE_UPDATE
                        Always updates as long as there is a release
  -f [FAST_MESSAGES], --fast-messages [FAST_MESSAGES]
                        No delay when printing out messages to the console. This is always on with --check
  ```

- Examples:
   -  Testing on a linux machine
      - **Basic update**: typically you want to run without using `/dev/console` and probably not using `/storage`.
        - Ex: `python3 ./packages/351elec/sources/scripts/get-release.py --update-dir ~/updates --console /dev/stderr`
      - **Check only** `python3 ./get-release.py --update-dir ~/updates --console /dev/stderr --check True`
      - **Different repository to look for releases** `python3 ./get-release.py --update-dir ~/updates --console /dev/stderr --org pkegg`
      - **More logging** (and redirect the console to a file) `python3 ./get-release.py  --update-dir ~/updates --console ~/updates/.console --log-level DEBUG`
    - Testing on RG351 device (via ssh - if you don't want to make a full 'build')
      - `/usr/bin/updatecheck` - this will run the check to see if an update is available.  Returns `no` or the version.
      - `/usr/bin/351elec-upgrade` - this is the actual script to upgrade.
        - **NOTE**: Both of the above scripts can be run in another directory (`/storage/`) by copying them and `/usr/bin/get-release.py` into the same directory.  That would allow changing for debugging on the device, etc.
      - `python3 /usr/bin/get-release.py` will also work as above, etc.  You may need to manually shutdown emulation station with `systemctl stop emustation` in order to see what is printed to the console (`/dev/console`)
 
# Testing on the RG351
**DEV: How to test before we have real releases**.  There are new developer options under 'Emulationstation Settings' -> 'Developer' -> 'Updates' for "GITHUB ORG" and "GITHUB REPO".  To test with my PR updates, you can set GITHUB ORG to `pkegg`.

**DEV: How to force an update**: You can force an update by setting the developer setting "FORCE UPDATES" this will disable the check for the version on the device and always detect an update (and run an update if one exists).  You can use this to 'upgrade' back to the current release (remove GITHUB_ORG=`pkegg` so it will downgrade you to 20210603 from 351ELEC). 

Here’s what has been tested.  

- Basic update on V and M (mostly tested on the V, but M worked effectively the same for all this stuff)
- Ensure ‘no update available’ works properly and does not allow user to attempt update.
- Ensure ‘release’ update channel downloads proper update (from pkegg)
- Ensure ‘beta’ update channel downloads prerelease (if newer than release) - from pkegg
- Ensure “check for updates” pop up works, popping up about 10 seconds after ES starts.
- Ensure partial failed download is removed if sha256 does not match
- Ensure download fails w\o internet
- Ensure ‘channel’ stays on reboot (removed code that was setting to daily on reboot)
- `Force Updates` (developer option) works to skip the version check of what is on the device and always detect an upgrade if one exists in github.  I used this to upgrade back to the latest 351ELEC release when no GITHUB ORG is set.

### CODE REVIEW

File: packages/ui/351elec-emulationstation/package.mk [hunk 1]
```diff
@@ -3,12 +3,12 @@
...
 PKG_NAME="351elec-emulationstation"
-PKG_VERSION="58a5b5701ea07e27993bca4b569aee4ba0e3de39"
-PKG_GIT_CLONE_BRANCH="main"
+PKG_VERSION="b8f9f43441016079d8958cb86bf416f43806d749"
+PKG_GIT_CLONE_BRANCH="reenable-updates"
 PKG_REV="1"
 PKG_ARCH="any"
 PKG_LICENSE="GPL"
-PKG_SITE="https://github.com/351ELEC/351elec-emulationstation"
+PKG_SITE="https://github.com/pkegg/351elec-emulationstation"
```
Comment (takano32) [inline packages/ui/351elec-emulationstation/package.mk:11]: It's need to merge to upstream?

**Comment by pkegg** [comment]
Removed **pkegg** repository reference from emulationstation.  Switched to using the latest 351elec-emulationstation version as dhwz merged my changes there. :pray:

**Review by Cebion — APPROVED** [review]
Very well documented. The userbase will greatly appreciate this change!
Thank you so much!

**Review by dhwz — APPROVED** [review]
